channels: log contents of invalid message

justinmk opened the pull request:
~~This just a first step, it seems to be kind of useless, because (I guess) `log_msg_close .. msgpack_object_print` has nothing useful to print if the message was invalid.~~

- [x] save raw stream bytes for error-reporting
- [x] test `$NVIM_LOG_FILE` contents
- [ ] test for https://github.com/neovim/neovim/pull/8883#discussion_r211848065
- [ ] test currently causes `abort()` :
  ```
  :let c = sockconnect('pipe', v:servername, {'rpc':1})
  :echo chansend(c, "foo \0 \n\t ??  \0 \n \r")        
  ```
- [ ] is `channel_create_event` done for all channel types (e.g. `--embed` ch 1)?

justinmk committed:
RPC: log invalid message

Usually we dump the message contents only for DEBUG-level logging. But
INVALID message is unusual, so we should always dump it (helps with bug
reports, for example).

Caveat: the message contents of an invalid message aren't actually that
useful because msgpack-c probably failed to parse it.  So in practice,
this change only gives us the message id and type.

justinmk committed:
channels: log "loopback" status

justinmk commented on src/nvim/channel.c:
```diff
@@ -372,10 +376,11 @@ uint64_t channel_connect(bool tcp, const char *address,
   Channel *channel;
+  bool loopback = false;
 
   if (!tcp && rpc) {
```
TODO: writing an invalid message to a bytes (not RPC) "loopback" channel closes the stream, but not the channel, so `chansend()` still claims success but internally there's a SIGPIPE.

```vim
:let c = sockconnect('pipe', v:servername)
:echo chansend(c, "foo\n")
:echo chansend(c, "foo\n")
:edit $NVIM_LOG_FILE
```

bfredl replied:
Hmm chansend should always check the stream, not the entire channel. Formally this is a pair of "normal" channels without any extra "loopback" logic, so I don't see what should go wrong here.

bfredl commented on src/nvim/channel.c:
```diff
@@ -185,7 +185,7 @@ static Channel *channel_alloc(ChannelStreamType type)
 }
 
-void channel_create_event(Channel *chan, const char *ext_source)
+void channel_create_event(Channel *chan, const char *ext_source, bool loopback)
```
This info is already available as `chan->streamtype == kChannelStreamInternal`

justinmk committed:
rbuffer.c: doc

justinmk committed:
RPC: mention message-id in out-of-sync error

justinmk committed:
RPC: log (partial) contents of invalid message

If msgpack or msgpack-RPC validation fails, it is very helpful to see
a hint in the logs about the contents of the stream. log_client_msg() is
useless for that purpose because msgpack_object_print() expects a valid
msgpack object.

justinmk committed:
RPC: parse_msgpack(): return early on invalid message

Otherwise msgpack_unpacker_next() iterates through the stream and
reports "invalid message" for each byte.

justinmk commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -246,25 +259,31 @@ static void parse_msgpack(Channel *channel)
+    if (!handle_request(channel, &unpacked.data)) {
+      // Bail out from this event loop iteration
+      return false;
+    }
```
We already bail out early on line 280. Without this change, `msgpack_unpacker_next` keeps iterating through every byte, then `handle_request` fails again and again.

justinmk committed:
fixup: add test
